tests/test_slurm_system.py

```diff
@@ -155,3 +155,26 @@ def test_update_node_states_with_mocked_outputs(mock_get_sinfo, mock_get_squeue,
 def test_parse_node_list(node_list: str, expected_parsed_node_list: List[str], slurm_system):
     parsed_node_list = slurm_system.parse_node_list(node_list)
     assert parsed_node_list == expected_parsed_node_list
+
+
+def test_get_available_nodes_from_group_max_avail(slurm_system):
+    group_name = "backup"
+    partition_name = "backup"
+
+    slurm_system.groups = {
+        "backup": {
+            group_name: [
+                SlurmNode(name="node01", partition=partition_name, state=SlurmNodeState.IDLE),
+                SlurmNode(name="node02", partition=partition_name, state=SlurmNodeState.COMPLETING),
+                SlurmNode(name="node03", partition=partition_name, state=SlurmNodeState.IDLE),
+                SlurmNode(name="node04", partition=partition_name, state=SlurmNodeState.DOWN),
+            ]
+        }
+    }
+
+    available_nodes = slurm_system.get_available_nodes_from_group(partition_name, group_name, "max_avail")
```

Comment:
We can focus on testing `allocate_nodes(...)` as it has less dependencies. It should be easier to test different options of `number_of_nodes`.

Existing test is also useful and make sense. Please extend it to test non-`"max_avail"` cases.